// vite-plugin-stylelint/dist/index.cjs
'use strict';

const node_path = require('node:path');
const node_url = require('node:url');
const node_worker_threads = require('node:worker_threads');
const debugWrap = require('debug');
const utils = require('./shared/vite-plugin-stylelint.DtpuXS4O.cjs');
require('@rollup/pluginutils');

var _documentCurrentScript = typeof document !== 'undefined' ? document.currentScript : null;
function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e.default : e; }

const debugWrap__default = /*#__PURE__*/_interopDefaultCompat(debugWrap);

const debug = debugWrap__default(utils.PLUGIN_NAME);
const __filename$1 = node_url.fileURLToPath((typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.tagName.toUpperCase() === 'SCRIPT' && _documentCurrentScript.src || new URL('index.cjs', document.baseURI).href)));
const __dirname$1 = node_path.dirname(__filename$1);
const ext = node_path.extname(__filename$1);
function StylelintPlugin(userOptions = {}) {
  const options = utils.getOptions(userOptions);
  let worker;
  const filter = utils.getFilter(options);
  let stylelintInstance;
  let formatter;
  return {
    name: utils.PLUGIN_NAME,
    apply(config, { command }) {
      debug("==== apply hook ====");
      if (config.mode === "test" || process.env.VITEST) return options.test;
      const shouldApply = command === "serve" && options.dev || command === "build" && options.build;
      debug(`should apply this plugin: ${shouldApply}`);
      return shouldApply;
    },
    async buildStart() {
      debug("==== buildStart hook ====");
      if (options.lintInWorker) {
        if (worker) return;
        debug("Initialize worker");
        worker = new node_worker_threads.Worker(node_path.resolve(__dirname$1, `worker${ext}`), {
          workerData: { options }
        });
        return;
      }
      debug("Initial Stylelint");
      const result = await utils.initializeStylelint(options);
      stylelintInstance = result.stylelintInstance;
      formatter = result.formatter;
      if (options.lintOnStart) {
        debug("Lint on start");
        await utils.lintFiles(
          {
            files: options.include,
            stylelintInstance,
            formatter,
            options
          },
          this
          // use buildStart hook context
        );
      }
    },
    async transform(_, id) {
      debug("==== transform hook ====");
      debug(`id: ${id}`);
      const watchIds = this.getWatchFiles();
      debug(`watchIds: ${watchIds}`);
      const ids = [id, ...watchIds];
      debug(`ids: ${ids}`);
      if (worker) {
        for (const id2 of ids) {
          worker.postMessage(id2);
        }
        return;
      }
      const filteredIds = ids.filter((id2) => !utils.shouldIgnoreModule(id2, filter));
      debug(`filteredIds: ${filteredIds}`);
      const shouldIgnore = filteredIds.length === 0;
      debug(`should ignore: ${shouldIgnore}`);
      if (shouldIgnore) return;
      const filePaths = filteredIds.map((id2) => utils.getFilePath(id2));
      return await utils.lintFiles(
        {
          files: options.lintDirtyOnly ? filePaths : options.include,
          stylelintInstance,
          formatter,
          options
        },
        this
        // use transform hook context
      );
    },
    async buildEnd() {
      debug("==== buildEnd hook ====");
      if (worker) await worker.terminate();
    }
  };
}

module.exports = StylelintPlugin;
